### ISSUE TITLE
MVP QA Dashboard Hosted in DOMO

### ISSUE THREAD

**Issue by pjhill** [department-of-veterans-affairs/va.gov-team#22754]
## Issue Description
Create a MVP dashboard that collects data from the results of test methodologies on the Platform.
* NYC/Istanbul code coverage report for unit tests
* MochaAwesome for unit tests results
* MochaAwesome for Cypress test results (not coverage)
* SimpleCov for Rspec code coverage
* [JSON formatter for Rspec](https://relishapp.com/rspec/rspec-core/docs/formatters/json-formatter) (potentially)
* TestRail results (will need to explore for existing tools or a custom solution for retrieving and formatting data from the API)
* PACT???
* Locust tests

### Mapping Raw Data Stream to Metrics
#### Manual Functional Testing
| Metric                                                                                                          | Metric Group       | Automated | Raw Data Source                                                                                                                                                                                    | Notes                                                                                                                                                                                                                                                                                |
|-----------------------------------------------------------------------------------------------------------------|--------------------|-----------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| VFS Teams are satisfied with their ability to verify functional requirements                                    | Ease of Use        | N         |                                                                                                                                                                                                    |                                                                                                                                                                                                                                                                                      |
| DONE, LAST 30 DAYS, LAST 7 DAYS - % of active VFS Teams with completed test plans in TestRail (should be 100% because of collab cycle guidelines) | Service Completion | Y         | https://www.gurock.com/testrail/docs/api/reference/plans#getplans                                                                                                                                  | Consider using :is_completed to understand whether the projectID contains a completed test plan which would mean that the VFS team is meeting some success in using TestRail                                                                                                         |
| DONE, LAST 30 DAYS, LAST 7 DAYS - # of users actively engaged in the use of TestRail                                                              | Findability        | Y         | https://www.gurock.com/testrail/docs/api/reference/cases#getcase                                                                                                                                   | Examine which users have created a test case within  in order to determine whether a user has been active                                                                                                                                                                            |
| DONE, LAST 30 DAYS, LAST 7 DAYS (IS THIS A DUPLICATE OF THE PREVIOUS ONE?) - % of users with account activity in the last month                                                              | Service Completion | Y         | https://www.gurock.com/testrail/docs/api/reference/cases#getcase                                                                                                                                   | Examine which users have created a test case within  in order to determine whether a user has been active and compare against total Using an administrator API key means that we don't have to pass project_id and we could get a list of all users regardless of associated project |
| % of users with a positive disposition toward TestRail                                                          | Satisfaction       | N         |                                                                                                                                                                                                    |                                                                                                                                                                                                                                                                                      |
| DONE, LAST 30 DAYS, LAST 7 DAYS - % of teams with completed test plans                                                                            | North Star         | Y         | https://www.gurock.com/testrail/docs/api/reference/plans#getplans                                                                                                                                  |                                                                                                                                                                                                                                                                                      |
| % of teams with defects linked from a failing test case in TestRail                                             | Service Completion | Y         | https://www.gurock.com/testrail/docs/api/reference/tests#gettest https://www.gurock.com/testrail/docs/api/reference/statuses https://www.gurock.com/testrail/docs/api/reference/results#getresults | Retrieve all tests from a project in order to get a set of test objects in a project, examine the status of the test objects to get the subset that failed, and use getresults to determine whether the failed tests have defects associated                                         |
| % of teams with test runs across the 5 most used products on va.gov                                             | Service Completion | Y         | Call GA to determine the 5 most used products on the platform and use a lookup table to call TestRail API and retrieve test runs for the given project                                             |                                                                                                                                                                                                                                                                                      |
| VFS Team survey reported satisfaction                                                                           | Satisfaction       | N         |                                                                                                                                                                                                    |                                                                                                                                                                                                                                                                                      |

#### E2E Testing
| Metric                                                                                                | Metric Group       | Automated | VFS Team vs. VSP Personnel | Raw Data Source                                                                                       | Notes                                                                                                                                                                                                           |
|-------------------------------------------------------------------------------------------------------|--------------------|-----------|----------------------------|-------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| % of VFS teams creating E2E tests                                                                     | Service Completion | N         | Both                       | -                                                                                                     | -                                                                                                                                                                                                               |
| # of minutes the E2E test suite takes to execute in CI                                                | North Star         | Y         | Both                       | https://github.com/jenkinsci/pipeline-stage-view-plugin/tree/master/rest-api#get-jobjob-namewfapiruns | Retrieve `durationMillis` from json for name=Integration                                                                                                                                                        |
| % VFS Teams creating Cypress test specs for new product initiatives                                   | Service Completion | Y         | VSP                        | https://docs.github.com/en/rest/reference/search#search-code                                          | Use GH's search code endpoint to discover all cypress spec files and where they are stored. Compare against a master list of products                                                                           |
| % of new product FE PRs that include an E2E test spec                                                 | Service Completion | Y         | VSP                        | https://docs.github.com/en/rest/reference/search#search-issues-and-pull-requests                      | Use GH's search pull request endpoint to discover % of PRs that contain changes to Cypress specs                                                                                                                |
| % of tests employing best practice methods (viewport tester, form tester, e2e keyboard helper, etc.)  | Ease of Use        | Y         | VSP                        | https://docs.github.com/en/rest/reference/search#search-code                                          | Use GH's search code endpoint to discover all cypress spec files and examine their contents for keywords related to best practice test design                                                                   |
| # CI build failures due to flaky test execution                                                       | Ease of Use        | N         | VSP                        | -                                                                                                     | I'm not sure how to retrieve this information. It would involve parsing of detailed logs / results. Perhaps we can consume the Mochawesome report to retrieve this info?                                        |
| # of VFS teams that release with E2E tests in place                                                   | Service Completion | Y         | VSP                        | https://docs.github.com/en/rest/reference/search#search-code                                          | Use GH's search code endpoint to discover all cypress spec files and file path. Use file path to compare to a master list of projects that are released.                                                        |
| # of VFS Teams posing questions regarding E2E testing into platform support or testing tools channels | Ease of Use        | Y         | VSP                        | https://docs.github.com/en/rest/reference/search#search-issues-and-pull-requests                      | We currently collect support issues into a repo of issues created by the support bot. Use the search issues functionality to sift through the support bot repository for issues related to E2E testing keywords |
| # of VFS teams reporting flaky e2e test issues                                                        | Ease of Use        | Y         | VSP                        | https://docs.github.com/en/rest/reference/search#search-issues-and-pull-requests                      | We currently collect support issues into a repo of issues created by the support bot. Use the search issues functionality to sift through the support bot repository for issues related to E2E testing keywords |
| # of minutes the E2E test suite takes to execute locally                                              | Ease of Use        | N         | Both                       | -                                                                                                     | This can be entered into Domo / BigQuery on a schedule with some reference "local" machines that members of the test team operate.                                                                              |

#### Accessibility Testing
| Metric                                                            | Metric Group       | Automated | VFS Team vs. VSP Personnel | Raw Data Source | Notes                                                                                                                  |
|-------------------------------------------------------------------|--------------------|-----------|----------------------------|-----------------|------------------------------------------------------------------------------------------------------------------------|
| % of teams using Keyboard E2E tests                               | Service Completion | Y         |                            |                 |                                                                                                                        |
| % of teams who’s E2E tests contain aXe checks                     | Service Completion | Y         |                            |                 |                                                                                                                        |
| Counts of accessibility issues over time, per product and overall | North Star         | Y         |                            |                 | Cross reference team labels with a11y defect labels                                                                    |
| Issues by severity over time, per product and overall             | Service Completion | Y         |                            |                 | Cross reference team labels with a11y defect severity labels                                                           |
| Issue type over time, per product and overall                     | Service Completion | Y         |                            |                 | Issue type is describe in the issue title (IIRC), so cross reference team labels with extracted and parsed issue title |
| Time to fix issues - time to open vs. time to close               | Ease of Use        | Y         |                            |                 | As indicated, compare datetime for open to datetime for close                                                          |
| Tracking audit time from start to completion                      | Ease of Use        | Y         |                            |                 |                                                                                                                        |
| Count of open and closed accessibility issues                     | Service Completion |           |                            |                 |                                                                                                                        |
| Count of accessibility issues by defect severity                  | Service Completion |           |                            |                 |                                                                                                                        |
| Count of accessibility issues by type                             | Service Completion |           |                            |                 |                                                                                                                        |
| Average time to fix issues (All Teams)                            | Ease of Use        |           |                            |                 |                                                                                                                        |
| Average time to fix issues (Product Team)                         | Ease of Use        |           |                            |                 |                                                                                                                        |
| Teams with number of issues above the median                      | Overall            |           |                            |                 |                                                                                                                        |
| Issues by type rising % faster than historical average            | Ease of Use        |           |                            |                 |                                                                                                                        |
| Issues by defect severity rising % faster than historical average | Service Completion |           |                            |                 |                                                                                                                        |
| Issues open longer than average number of days                    | Service Completion |           |                            |                 |                                                                                                                        |

---
## Tasks
- [ ] Work out hosting solution
- [ ] Schedule job to generate and retrieve reports (Once a day?) (GitHub Actions?)
- [ ] CRUD app to manipulate the list of manually collected metrics for display visually
- [ ] Create a presentation using the collected metrics from the CRUD app
- [ ] High level navigation to collected reports for more details on results and coverage in various test methodologies
- [ ] NYC/Istanbul summary JSON file parser to separate the references files into logical apps and aggregate the coverage score from all test files owned by an app and display that app's aggregate score

## Acceptance Criteria
- [ ] As a VSP stakeholder, I am able to visit the dashboard and quickly understand the overall QA health of the platform.
- [ ] As a VSP stakeholder, I am able to visit the dashboard and dig into the details of the QA health for a particular VFS team's product.
- [ ] As a VFS team stakeholder, I am able to visit the dashboard and dig into the details of the QA health for my product.

**Comment by holdenhinkle** [comment]
Fleshing out the idea a little more:

## VSP Testing Suite Metrics
- The metrics on the 'VSP Testing Suite' sheet in [this spreadsheet](https://docs.google.com/spreadsheets/d/1UWbOr9avjfJJPGaG4OHJY5MNd-zGkMvESce7dWPmIl8/edit#gid=0&range=A1) will be updated once a month
- Create forms for the metrics in each test category listed on the sheet
- Alternatively, upload a `.csv` file to update the metrics all at once
- Each test category has its own page that displays graphs showing changes over time

## Metrics For Each Test Type
All BE and FE tests will run once a day, and include the following:

**BE Tests**
- RSpec Unit Tests
  - Test Results
    - A `JSON` test results file includes details about the test run
  - Code Coverage Results
    - A SimpleCov `HTML` report is generated for viewing code coverage results
    - SimpleCov also generates a `JSON` file that includes details about the test run (**this must be verified)

**FE Tests**
- Mocha Unit Tests
  - Test Results
    - A Mochawesome `HTML` report is generated for viewing test results
    - Mochawesome also generates a `JSON` file that includes details about the test run
  - Code Coverage Results
    - A NYC/Istanbul `HTML` report is generated for viewing code coverage results
    - NYC/Istanbul also generates a `JSON` file that includes details about the test run
- Cypress E2E Tests
  - Test Results
    - A Mochawesome `HTML` report is generated for viewing test results
    - Mochawesome also generates a `JSON` file that includes details about the test run

**Other Tests**
- Contract (Pact)
- ~~Load (Locust)~~

### JSON Files
`JSON` files from each test type (ie RSpec, Cypress, etc.) will be ingested and we'll save the following data to track changes over time, both for overall test results, and test results by app:
- Test Results
  - Number of tests
  - Num Passed
  - Num Failed
  - Num Failed
- Code Coverage Results
  - Total Files
  - Relevant Lines
  - Lines Covered
  - Lines Missed

These Metrics can be displayed by graphs. Additionally, we can calculate the overall health of VA.gov and each application in it from a testing point of view. Score cards can be created and we can also track these scores over time.

### HTML Reports
Finally, the Mochawesome, SimpleCov, NYC/Istanbul, and other(?) reports can be updated once a day and be presented within iFrames in the app (there's no sense in recreating these useful reports).

## Summary
This simple Test Dashboard will display:
- Updated Monthly:
  - [VSP Testing Suite Metrics](https://docs.google.com/spreadsheets/d/1UWbOr9avjfJJPGaG4OHJY5MNd-zGkMvESce7dWPmIl8/edit#gid=0&range=A1) over time
- Update Once A Day:
  - Overall test metrics
  - Test metrics by app
  - Overall score
  - Scores for each app
  - Detailed test reports

**Comment by holdenhinkle** [comment]
Waiting on access to DOMO.

**Comment by holdenhinkle** [comment]
Submitted request to #vsp-analytics for BigQuery service credentials - https://dsva.slack.com/archives/CKCRMLGAC/p1620069510243000

**Comment by holdenhinkle** [comment]
Some updates:

I submitted this ticket to ops the other day - https://github.com/department-of-veterans-affairs/va.gov-team/issues/24247

Demian contacted me in a Slack DM and gave me what I need.

Demian Ginther  3:34 PM
to access a variable in a Github Action, you can use something like this:
```
    - name: Configure AWS Credentials
      uses: aws-actions/configure-aws-credentials@v1
      with:
        aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
        aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
        aws-region: "us-gov-west-1"
    - uses: marvinpinto/action-inject-ssm-secrets@v1.1.1
      with:
        ssm_parameter: /devops/dockerhub/vavfsbot_password
        env_variable_name: DOCKERHUB_PASSWORD
```

Demian Ginther  3:35 PM
and then you would use
```
    - name: Build AMI
      run: whatever you are running
      env:
        DOCKERHUB_PASSWORD: ${{ env.DOCKERHUB_PASSWORD}}
```

Demian Ginther  3:36 PM
you dont want to grab it in the js file, I would imagine. You just want to inject it into the environment your js runs in
and then you can grab that from your environment in your js


Demian Ginther  5:05 PM
and there is a path in Parameter Store: `/dsva-vagov/testing-team/`

Demian Ginther  5:05 PM
you should have access to write there

**Comment by holdenhinkle** [comment]
Continuing...

I created this repo the other day for the GitHub Action that will run a scheduled event to grab data from TestRail, build the metrics we need, and post it to BigQuery - https://github.com/department-of-veterans-affairs/testing-tools-team-dashboard-testrail-data

The action will run once a day at midnight.

Yesterday I wrote a script that grabs the TestRail data and creates the following metrics we identified:

```
{
  totalUsers: 23,
  totalVfsTeams: 22,
  totalActiveVfsTeams: 10,
  percentActiveVfsTeams: 45.45454545454545,
  totalActiveVfsUsers: 5,
  percentActiveVfsUsers: 21.73913043478261
}
```

I plan on getting the GitHub action working.

I'm still waiting for the BigQuery service credentials from the Analytics team. I'll be blocked at the end of the day without them.

**Comment by holdenhinkle** [comment]
I got the scheduled GitHub action working - https://github.com/department-of-veterans-affairs/testing-tools-team-dashboard-testrail-data/actions

The job is scheduled to run at 12am, UTC each day.

It turns out that there's often a delay between when a job is scheduled to run and when it actually runs. Here are a few posts I found about the issue:
- https://github.community/t/scheduled-action-running-consistently-late/138025
- https://github.community/t/no-assurance-on-scheduled-jobs/133753

**Comment by pjhill** [comment]
> I got the scheduled GitHub action working - https://github.com/department-of-veterans-affairs/testing-tools-team-dashboard-testrail-data/actions
> 
> The job is scheduled to run at 12am, UTC each day.
> 
> It turns out that there's often a delay between when a job is scheduled to run and when it actually runs. Here are a few posts I found about the issue:
> 
> * https://github.community/t/scheduled-action-running-consistently-late/138025
> * https://github.community/t/no-assurance-on-scheduled-jobs/133753
> 
> Starting to write unit tests for this action.
> 
> Then I'll need the BQ credentials!

I have the BQ credentials available, and I can Keybase them to you.

**Comment by pjhill** [comment]
Additional TestRail metrics to consider --

How many test cases in a product's test plan discovered a defect?
How many defects were discovered in a test run for a given product?
How many test cases platform-wide have discovered defects? Percentage?
What percent of test cases have never discovered a defect?

**Comment by pjhill** [comment]
Visual design thoughts -- 
![image.png](https://images.zenhubusercontent.com/5d2633e0a6db635dbcbaf8d5/967202a1-4fdb-4c0b-959e-66bf82abf18c)

* Pie chart of distribution of defects discovered across product teams
* Bar chart rate of defect discovery over time
* Consider that we want to evaluate product health in the future. This will probably require an interpretation layer that sits between this higher level view and the more raw data coming in from our tool APIs.

**Comment by pjhill** [comment]
The MPV is done -- future efforts will be under post-MVP initiatives!
